### PR TITLE
Unstructured plots

### PR DESCRIPTION
**Pull request by doutriaux1**
@dlonie @chaosphere2112 please review
dashboard: https://open.cdash.org/buildSummary.php?buildid=3780331

### CODE REVIEW

**Comment by doutriaux1** [comment]
goes with: https://github.com/UV-CDAT/uvcdat-testdata/pull/35

**Comment by doutriaux1** [comment]
@ThomasMaxwell can you please try it too? I'm a bit worried it breaks your unstructured grids! Your tests pass but I would feel better if you would try your unstructured grids plotting on this.

**Comment by ThomasMaxwell** [comment]
I don’t see any effect of this branch upon the vcs3D unstructured grid plots.

— Tom

File: Packages/vcs/Lib/Canvas.py [hunk 1]
```diff
@@ -2759,16 +2764,20 @@ def __plot (self, arglist, keyargs):
...
             except Exception,err:
                 arglist[4]='default'
+        elif g0 is not None and (arglist[0] is not None and isinstance(arglist[0],cdms2.avariable.AbstractVariable) and not isinstance(g0,cdms2.grid.AbstractRectGrid)) and arglist[3] in ["boxfill","default"] and arglist[4]=="default":
+          arglist[3]="meshfill"
 
 ##                         arglist[4]=copy_mthd.name
-
+        # Ok let's check for meshfill needed
+        if g0 is not None and (arglist[0] is not None and isinstance(arglist[0],cdms2.avariable.AbstractVariable) and not isinstance(arglist[0].getGrid(),cdms2.grid.AbstractRectGrid)) and arglist[3] not in ["meshfill",]:
+          raise RuntimeError("You are attempting to plot unstructured grid with a method that is not meshfill")
```
Comment (doutriaux1) [inline Packages/vcs/Lib/Canvas.py]: @ThomasMaxwell thanks for checking. Specifically that is the area I was worried about. Glad it didn't break your stuff.

File: Packages/vcs/Lib/Canvas.py [hunk 2]
```diff
@@ -2708,9 +2710,12 @@ def __plot (self, arglist, keyargs):
             if float(doratio[:-1])==0.: doratio='0'
 
         ## Check for curvilinear grids, and wrap options !
+        if arglist[0] is not None:
+            g0=arglist[0].getGrid()
```
Comment (aashish24) [inline Packages/vcs/Lib/Canvas.py]: what is g0 represents here (why 0?) Can we use more descritive variable names?

**Comment by aashish24** [comment]
@dlonie can you approve this branch (just verify my changes). The tests are passing for me. 

@doutriaux1 I cannot verify if the baseline looks correct. Can you double check on that please?

File: testing/vcs/test_plot_unstructured_via_boxfill.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+import checkimage
+
+f=cdms2.open(os.path.join(vcs.prefix,"sample_data","sampleCurveGrid4.nc"))
+
+s=f("sample")
+x=vcs.init()
+x.drawlogooff()
+x.setbgoutputdimensions(1200,1091,units="pixels")
+
+import vcs
```
Comment (dlonie) [inline testing/vcs/test_plot_unstructured_via_boxfill.py]: These imports (and the checkimage/open below) don't need to be repeated. Accidental copy/paste?

File: testing/vcs/test_plot_unstructured_via_boxfill.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+src=sys.argv[1]
+
+x.png(fnm)
+
+print "fnm:",fnm
+print "src:",src
+ret = checkimage.check_result_image(fnm,src,checkimage.defaultThreshold)
+sys.exit(ret)
+
+fnm = "test_plot_unstructured_via_boxfill.png"
```
Comment (dlonie) [inline testing/vcs/test_plot_unstructured_via_boxfill.py]: Same here, is this repeated intentionally?

**Comment by dlonie** [comment]
I don't see any functional issues, but the repeated lines of code in the test are confusing.

**Comment by aashish24** [comment]
good catch. I took care of it. Please have a look again

**Comment by aashish24** [comment]
Grr...Okay..pushed a fix again.

**Comment by aashish24** [comment]
@dlonie I guess it should be all good now.

**Comment by dlonie** [comment]
+1. You beat me to the missing newline ;)

**Comment by dlonie** [comment]
Can't merge for conflicts, though.

**Comment by aashish24** [comment]
@dlonie wait before you review it.

**Comment by aashish24** [comment]
Okay now it should be good to review.

File: testing/vcs/CMakeLists.txt [hunk 5]
```diff
@@ -266,6 +266,15 @@ cdat_add_test(vcs_test_taylor_2_quads
 
   # These test actually plot things need sample data
   if (CDAT_DOWNLOAD_SAMPLE_DATA)
+    cdat_add_test(test_vcs_plot_unstructured_via_non_meshfill_or_default_fails
```
Comment (dlonie) [inline testing/vcs/CMakeLists.txt]: This file seems to be missing from the branch.

Comment (dlonie) [inline testing/vcs/CMakeLists.txt]: The test script, that is.

File: testing/vcs/test_vcs_dump_json.json [hunk 6]
```diff
@@ -112,8 +112,8 @@
    "datawc_x2": 1e+20, 
    "datawc_y1": 1e+20, 
    "datawc_y2": 1e+20, 
-   "ext_1": false, 
-   "ext_2": false, 
+   "ext_1": false,
```
Comment (dlonie) [inline testing/vcs/test_vcs_dump_json.json]: This is breaking the `vcs_test_dump_json` test:

```
test 62
    Start 62: vcs_test_dump_json

62: Test command: /ssd/src/uvcdat-build/install/bin/python "/ssd/src/uvcdat/testing/vcs/test_dump_json.py" "/ssd/src/uvcdat/testing/vcs/test_vcs_dump_json.json"
62: Test timeout computed to be: 1500
62: Traceback (most recent call last):
62:   File "/ssd/src/uvcdat/testing/vcs/test_dump_json.py", line 31, in <module>
62:     assert(filecmp.cmp("test_vcs_dump_json.json",src))
62: AssertionError
1/1 Test #62: vcs_test_dump_json ...............***Failed    1.05 sec
```

**Comment by aashish24** [comment]
Ah, okay, I don't know why I didn't see that in my ctest. Let me have a look at it.

**Comment by aashish24** [comment]
@dlonie Should be fixed now.

**Comment by dlonie** [comment]
Tests pass, change looks good \o/

**Pull request merged**